src/Compiler/TypedTree/TypedTreeOps.fs

```diff
@@ -10360,3 +10360,34 @@ let isFSharpExceptionTy g ty =
     | ValueSome tcref -> tcref.IsFSharpException
     | _ -> false
 
+let (|EmptyModuleOrNamespaces|_|) (moduleOrNamespaceContents: ModuleOrNamespaceContents) =
+    match moduleOrNamespaceContents with
+    | TMDefs(defs = defs) ->
+        let mdDefsLength =
+            defs
+            |> List.count (function
+                | ModuleOrNamespaceContents.TMDefRec _
+                | ModuleOrNamespaceContents.TMDefs _ -> true
+                | _ -> false)
+        
+        let emptyModuleOrNamespaces =
+            defs
+            |> List.choose (function
+                | ModuleOrNamespaceContents.TMDefRec _ as defRec
+                | ModuleOrNamespaceContents.TMDefs(defs = [ ModuleOrNamespaceContents.TMDefRec _ as defRec ]) ->
+                    match defRec with
+                    | TMDefRec(bindings = [ ModuleOrNamespaceBinding.Module(mspec, ModuleOrNamespaceContents.TMDefs(defs = defs)) ]) ->
+                        defs
+                        |> List.forall (function
+                            | ModuleOrNamespaceContents.TMDefOpens _
+                            | ModuleOrNamespaceContents.TMDefDo _ -> true
+                            | _ -> false)
+                        |> fun isEmpty -> if isEmpty then Some mspec else None
+                    | _ -> None
+                | _ -> None)
+
+        if mdDefsLength = emptyModuleOrNamespaces.Length then
+            Some emptyModuleOrNamespaces
+        else
+            None
+    | _ -> None
```

Comment:
Should we also check for `TMDefRec` here? Consider these (and maybe add to tests):

```fsharp
module rec Foobar

do ()
```


```fsharp
namespace rec Foobar

do () 
```

Reply:
```fsharp
module rec Foobar

do ()
```

currently produces

```fsharp
module Foobar

val ``doval@3`` : unit
```

This probably isn't correct but is outside my scope here.